keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,294 @@
+serves the resourceVersion="0" list requests from reflectors today.
+
+### Goals
+
+- Resolve the "stale read" problem (https://github.com/kubernetes/kubernetes/issues/59848)
+- Improve the scalability and performance of Kubernetes for Get and List requests, when the watch cache is enabled
+
+## Proposal
+
+### Leveraging the Progress Notify Mechanism
+
+Guard this by a `WatchCacheConsistentReads` feature gate.
+
+#### Alternative 1: Use WithProgressNotify to enable automatic watch updates
+
+Create etcd watches with `WithProgressNotify` enabled (available in all etcd 3.x versions).
+
+When `WithProgressNotify` is enabled on an etcd watch, etcd sends progress
+events to the watch automatically. By default etcd sends progress events every
+10 minutes, which is not frequent enough to be useful for our needs, so we will
+modify etcd to send them must more frequently, e.g. every 20ms.
+
+When an consistent read request is received and the watch cache is enabled:
+
+- Get the current revision from etcd for the resource type being served. The returned revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
+- Use the existing `waitUntilFreshAndBlock` function in the watch cache to wait briefly (20ms?) for the watch to catch up to the current revision.
+- If the block times out, skip the cache and serve the request directly from storage (etcd).
+
+To get the revsion we have two options: 
+
+- Use an etcd range request with `WithCount` enabled so etcd return only a count and revision 
+- Use an etcd range request against a known empty range with limit=1 as an additional guard (since etcd does not allow for limit=0)
+
+We will benchmark and select the most efficient of these options.
+
+#### Alternative 2: Use WatchProgressRequest to request watch updates when needed
+
+etcd 3.4+ provides a `WatchProgressRequest` request that can be made on a watch channel. When requested,
+etcd will send a progress event on that watch as soon as possible.
+
+When an consistent read request is received and the watch cache is enabled:
+- Get the current revision from etcd using a range read with limit=0, just like in alternative 1.
+- If the cache already has the current revision, serve the request from cache
+- If the current revision is not in the cache:
+  - Send a `WatchProgressRequest` to etcd on the watch channel that the watch cache is consuming.
+- Use the existing waitUntilFreshAndBlock function in the watch cache to wait for the watch to catch up to the current revision
+- If the block times out, skip the cache and serve the request directly from storage.
+
+This alternative requires using `WatchProgressRequest` which is only available in etcd 3.4+, and so would
+require we make the kube-apiserver aware of etcd's minor version, which is described in more detail later.
+
+#### Comparing the alternatives
+
+We will experiment with modifying etcd to automatically send progress events in
```

Comment:
Something to consider, for the priority and fairness effort we are going to be letting less critical requests wait in queues. Which means by the time the request gets to the watch handler, it may have sat in a queue for N seconds, like as many as 15.
* Maybe 20ms is more aggressive than necessary?
* Maybe the data served only needs a guarantee that it's at least as recent as when the request arrived at apiserver?

Like, what if once or twice a second apiserver did a consistent read, and stores a mapping from time to RV. Then, when handling a request, it looks at the request's arrival time. If it's old enough that we've got a subsequent time:RV value, then we don't have to make a separate trip to etcd and probably the cache won't have to wait any additional time before serving.

Otherwise, we could make a special request of etcd to get the global index (as in the design as currently written), or we could just wait for the next one.

Reply:
> Maybe the data served only needs a guarantee that it's at least as recent as when the request arrived at apiserver?

Yes, this is certainly true. What we're specifically guaranteeing is linearizable consistency. An intuitive way to think about this is: If some agent performs some write and gets a response from etcd confirming that write, a subsequent read (either by that agent, or another agent that is told that write read as been completed) should return data at a state no older than that write. Delays before serving the read and delays returning the result are expected (being a distributed system and all) and don't impact correctness.

> Like, what if once or twice a second apiserver did a consistent read, and stores a mapping from time to RV. Then, when handling a request, it looks at the request's arrival time. If it's old enough that we've got a subsequent time:RV value, then we don't have to make a separate trip to etcd and probably the cache won't have to wait any additional time before serving.

Yes, this works. The happensBefore chain provides the exact same guarantee. When read concurrency is high, somehow minimizing the "what's the current RV" requests to etcd would make a lot of sense. They could either be periodic, or another why I had thought about doing it is to accumulate requests that are received over some interval, and at the interval make one "what's the current RV" request to etcd, and then serve all the requests from cache.

I'll add it to the potential optimizations.